packages/whenable/src/whenable.js

```diff
@@ -0,0 +1,111 @@
+// @ts-check
+import { makePromiseKit } from '@endo/promise-kit';
+import { M } from '@endo/patterns';
+import { makeTagged } from '@endo/pass-style';
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareWhenableKits = zone => {
+  /** WeakMap<object, any> */
+  const whenableV0ToEphemeral = new WeakMap();
+
+  /**
+   * Get the current incarnation's promise kit associated with a whenableV0.
+   *
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @returns {import('@endo/promise-kit').PromiseKit<any>}
+   */
+  const findCurrentKit = whenableV0 => {
+    let pk = whenableV0ToEphemeral.get(whenableV0);
+    if (pk) {
+      return pk;
+    }
+
+    pk = makePromiseKit();
+    pk.promise.catch(() => {}); // silence unhandled rejection
+    whenableV0ToEphemeral.set(whenableV0, pk);
+    return pk;
+  };
+
+  /**
+   * @param {'resolve' | 'reject'} kind
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @param {unknown} value
+   */
+  const settle = (kind, whenableV0, value) => {
+    const kit = findCurrentKit(whenableV0);
+    const cb = kit[kind];
+    if (!cb) {
+      return;
+    }
+    whenableV0ToEphemeral.set(whenableV0, harden({ promise: kit.promise }));
+    cb(value);
+  };
+
+  const makeWhenableInternalsKit = zone.exoClassKit(
+    'WhenableInternalsKit',
+    {
+      whenableV0: M.interface('WhenableV0', {
+        shorten: M.call().returns(M.promise()),
+      }),
+      settler: M.interface('Settler', {
+        resolve: M.call().optional(M.any()).returns(),
+        reject: M.call().optional(M.any()).returns(),
+      }),
+    },
+    () => ({}),
```

Comment:
Note to revisit after further review: is it a problem that a whenable has no state and its promise is therefore forgotten upon upgrade? What happens if e.g. a durably-stored whenable is resolved in more than one incarnation?

Reply:
Yes, it is a problem.  I have a sketch that I will work on:

1. We should start from `this.state.value = undefined`
2. After `kit.resolve(p)` or `kit.reject(e)`, we should set `this.state.value = 'pending'` and run `watchPromise(kit.promise, this.facets.watcher)` to obtain the final value or the next whenable, even if it takes multiple incarnations.
3. The watcher would save `[isRejection, awaited]` in `this.state.value`.
4. The internals methods will be sensitive to these states instead of relying on any state from the promise kits.

Reply:
I fear we need that fixed before this PR is merged. Yes? Or am I missing something?